docs/feature_rawhid.md

```diff
@@ -0,0 +1,84 @@
+# Raw HID
+
+Raw HID allows for bidirectional communication between QMK and the host computer over an HID interface. This has many potential use cases, such as switching keymaps on the fly or changing RGB LED colors and modes.
+
+There are two main components to getting raw HID working with your keyboard.
+
+## Keyboard firmware
+
+The implementation is fairly straightforward for the firmware.
+In your `rules.mk` add:
+
+```make
+RAW_ENABLE = yes
+```
+
+In your `keymap.c` include `"raw_hid.h"` and implement the following:
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    // Your code goes here. data is the packet received from host.
+}
+```
+
+The `"raw_hid.h"` header also includes `void raw_hid_send(uint8_t *data, uint8_t length);`
+which allows sending packets from keyboard to host. It can be used for debugging when building your host by returning all data back to host to test communication.
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    raw_hid_send(data, length);
+}
+```
+
+All Raw HID packets by default have size `RAW_EPSIZE` which is `32 bytes`, the actual length is determined by `uint8_t length` both when sending and receiving.
```

Comment:
This part is a little complex. When calling `raw_hid_send()` from QMK the packet and length *must* be 32 bytes, regardless of the actual message size.
https://github.com/qmk/qmk_firmware/blob/3eb82e047079230e55d3aeff901c4d0dce69fb3a/tmk_core/protocol/lufa/lufa.c#L143-L147
`raw_hid_receive()` however is able to take variable length messages up to 32 bytes.

The ATSAM code appears to set a `RAW_EPSIZE` of 64 bytes, which also seems possible on AVR, so I'd like to see that upped (though it may have to go through the breaking changes process - luckily we have a ton of boards running VIA to test 😄).

Reply:
Ok I updated that, not sure if it's we're quite there though. take a look.